_drafts/firmware-code-size-2.md

```diff
@@ -0,0 +1,303 @@
+* `-O3`: optimize for performance at all cost, no matter the code size or compilation time impact.
+* `-Ofast` a.k.a. ludicrous mode: take off the guard rails, and disregard strict C standard compliance in the name of speed. This may lead to invalid programs.
+* `-Os` optimize for code size. This uses `O2` as the baseline, but disables some optimizations. For example, it will not [inline](https://en.wikipedia.org/wiki/Inline_expansion) code if that leads to a size increase.
+
+
+Compiled with GCC’s default (`-O0`), our example clocks in at a whooping 108KiB of code:
+
+```
+Linking build/ch.elf
+Creating build/ch.hex
+Creating build/ch.bin
+Creating build/ch.dmp
+
+   text	   data	    bss	    dec	    hex	filename
+ 111072	    296	  98008	 209376	  331e0	build/ch.elf
+Creating build/ch.list
+```
+
+Our baseline in the previous section uses ChibiOS’s default of `-O2`. You’ll have guessed by now that we can do better by using `-Os`:
+
+```
+Linking build/ch.elf
+Creating build/ch.hex
+Creating build/ch.bin
+Creating build/ch.dmp
+
+   text	   data	    bss	    dec	    hex	filename
+  62956	    232	  98072	 161260	  275ec	build/ch.elf
+Creating build/ch.list
+```
+
+61KiB! This is 47KiB smaller than the default GCC settings, and 10KiB smaller than the default project settings.
+
+### What if my program is performance sensitive?
+
+Parts of your program may be too performance-sensitive to compile them with `-Os`. At Pebble, this was the case for our text layout code, which was critical to hitting our target frame rate when scrolling through a notification. This is generally only true for a few very specific functions or files.
+
+GCC offers two methods to selectively tweak the optimization level: at the file level, or at the function level.
+
+If you want to set optimization level for a single function, you can use the `optimize` function attribute:
+```c
+void __attribute__((optimize("O3"))) fast_function(void) {
+    // ...
+}
+```
+
+For a whole file, you can use the `optimize` pragma:
+```c
+#pragma GCC push_options
+#pragma GCC optimize ("O3")
+
+/*
+ * Code that needs optimizing
+ */
+
+#pragma GCC pop_options
+```
+
+Note that the `push_options` and `pop_options` pragmas are needed to make sure the rest of the file is compiled with the regular options.
+
+## Linker Garbage Collection
+Now that we’ve set the appropriate optimization level, we turn our attention to dead code elimination. Most programs contain dead code: it may come from libraries you use partially, or test functions you left in the final program. 
+
+Since the compiler operates on one file at a time, it does not have enough context to decide whether a function is dead code or not.  Consider a library which exposes a function to add numbers in an array: since the files in the library do not call that function, the compiler could think it is not used when other parts of our program rely on it.
```

Comment:
MAYBE:

> since the files in the library do not call that function, the compiler could think it is not used when other parts of our program rely on it.

feels a little wordy but not sure if this is better:

while nothing in the library itself may call this function, the compiler has no way of knowing if other files are using it